lib/modules/manager/pipenv/extract.ts

```diff
@@ -95,14 +93,10 @@ function extractFromSection(
         // TODO #22198
         dep.managerData!.nestedVersion = nestedVersion;
       }
-      if (requirements.index) {
-        if (is.array(pipfile.source)) {
-          const source = pipfile.source.find(
-            (item) => item.name === requirements.index,
-          );
-          if (source) {
-            dep.registryUrls = [source.url];
-          }
+      if (sources && is.object(requirements) && requirements?.index) {
```

Comment:
```suggestion
      if (sources && is.object(requirements) && requirements.index) {
```
then the `?` isn't required